### PR TITLE
Simplify Generation of Proxy Factory Functions

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR simplifies the code-gen for the proxy factory functions (`ice_twoway`, `ice_secure`, etc.).

For the mapped proxy interfaces, we generate declarations, not implementations.
This was so repetitive, I rolled it into a much simpler for-loop.

For the mapped proxy classes, we no longer generate the implementations at all.
Instead, I added an intermediate class which implements everything instead.

Previously, the inheritance was: `_FooPrxI extends _ObectPrxI implements FooPrx`
Now, it is: `_FooPrxI extends _ObjectPrxIWrapper<FooPrx> implements FooPrx`
And `ObjectPrxIWrapper extends _ObjectPrxI`, so this change should be backwards compatible.

----

An alternative approach is to delete this wrapper class, and just generate the implementations directly on the proxy classes themselves. If you'd find that preferable to adding this intermediary class, feel free to voice your opinion.

### CODE REVIEW

File: cpp/src/slice2java/Gen.cpp [hunk 1]
```diff
@@ -4313,93 +4313,37 @@ Slice::Gen::ProxyVisitor::visitInterfaceDefEnd(const InterfaceDefPtr& p)
...
-
-    out << sp;
-    out << nl << "@Override";
-    out << nl << p->name() << "Prx ice_connectionId(String connectionId);";
-
-    out << sp;
-    out << nl << "@Override";
-    out << nl << p->name() << "Prx ice_fixed(com.zeroc.Ice.Connection connection);";
+    // Generate overrides for all the methods on `ObjectPrx` with covariant return types.
+    static const string objectPrxMethods[] = {
```
Comment (bernardnormier) [inline cpp/src/slice2java/Gen.cpp:4317]: I would use constexpr and string_view here.

File: java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxIWrapper.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
+// Copyright (c) ZeroC, Inc. All rights reserved.
+
+package com.zeroc.Ice;
+
+/**
+ * @hidden Provides overloads of the proxy factory methods with covariant return types so that
+ *     applications do not need to downcast to the derived proxy type.
+ */
+@SuppressWarnings("unchecked")
+public abstract class _ObjectPrxIWrapper<T extends ObjectPrx> extends _ObjectPrxI {
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxIWrapper.java:10]: This class doesn't wrap anything as far as I can tell. What about:
`ProxyFactoryMethods<T>` or `_ProxyFactoryMethods<T>`?

**Review by bernardnormier — APPROVED** [review]
It would be useful to update the Ice/inheritance test to make sure this syntax works correctly with inheritance.

**Review by pepone — APPROVED** [review]
Looks good!

File: cpp/src/slice2java/Gen.cpp (hunk 1 above)
Comment (InsertCreativityHere) [reply cpp/src/slice2java/Gen.cpp]: Fixed!

File: java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxIWrapper.java (hunk 2 above)
Comment (InsertCreativityHere) [reply java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxIWrapper.java]: Yes, you're correct. I changed the name (but kept the underscore to signal that it's _kind of_ internal.

**Comment by InsertCreativityHere** [comment]
I added a test case in https://github.com/zeroc-ice/ice/pull/2314/commits/98fc270e2c425e314547f35f8b8fc7ef8036b05d

**Pull request merged**